### DISCUSSION TITLE
Type definitions and specifications do not accept empty list syntax

### DISCUSSION THREAD

**Discussion by mjkalyan** [lfe/lfe#496] [Q&A] [answered]
Using either empty list shorthand, `()`/`[]`, in a `deftype` or `defspec` will cause the compiler to emit the `bad () type definition` error. The similar shorthand notations for the empty map `#m()` and tuple `#()` work just fine.

Here's an example:

```lisp
;; Fails to compile with:
;; bad () type definition
(deftype emptylist [])

;; Also:
;; bad () type definition
[defspec (test 1)
  (((integer))
   [])]
```

**Comment by rvirding** [comment, marked as answer]
OK, the reason for this is that `( )` is used for wrapping a type and is not usable as a literal empty list, `nil`. This can however be done using `(nil)` for the type empty list, maybe not the clearest but it does work. So you would write:

```
(deftype empty_list (nil))
```

Note that using `[ ]` is just some syntactic sugar and is always equivalent to `( )`. Always!

Note also in both Erlang and LFE `nil` is just a normal atom it has no special meaning as such. In fact in LFE it can also be used as a variable. LFE is closer to Erlang than to lisp wrt to data.

I need to work a bit more on the type documentation in lfe_types.txt.

**Comment by rvirding** [comment]
But I will admit that the syntax for type definitions and function specifications, especially the function specifications is very hard to read. I have been considering maybe using a map in the function specifications, e.g.

```
(defspec (a 1) 
  #M(argtypes [(number)] returntype (float))
  )
```
with an optional extra field `constraints`. This would be longer but I think easier to read.

**Comment by rvirding** [comment]
I just want to say that the optional syntax has been modified to:

```
(defspec (a 1) 
  #M(arg-types [(number)] ret-type (float))
  )
```

but the optional `constraints` is still the same. These have now been added in the latest commit to the branch. Note the branch has been rebased.
